keps/20200901-standard-crd.md

```diff
@@ -0,0 +1,254 @@
+
+    # Enum with values: 'Owner', 'Merge', or 'None'
+    # Default value of 'Owner'
+    # Owner creates the secret and sets .metadata.ownerReferences of the resource
+    # Merge does not create the secret, but merges in the data fields to the secret
+    # None does not create a secret (future use with injector)
+    creationPolicy: 'Merge'
+
+    # specify a blueprint for the resulting Kind=Secret
+    template:
+      type: kubernetes.io/dockerconfigjson # or TLS...
+
+      metadata:
+        annotations: {}
+        labels: {}
+
+      # use inline templates to construct your desired config file that contains your secret
+      data:
+        config.yml: |
+          endpoints:
+          - https://{{ .data.user }}:{{ .data.password }}@api.exmaple.com
+
+      # Uses an existing template from configmap
+      # secret is fetched, merged and templated within the referenced configMap data
+      # It does not update the configmap, it creates a secret with: data["alertmanager.yml"] = ...result...
+      templateFrom:
+      - configMap:
+          name: alertmanager
+          items:
+          - key: alertmanager.yaml
+
+  # data contains key/value pairs which correspond to the keys in the resulting secret
+  data:
+
+    # EXAMPLE 1: simple mapping
+    # one key from a store may hold multiple values
+    # we need a way to map the values to the frontend
+    # it is the responsibility of the store implementation to know how to extract a value
+    tls.crt:
+      key: /corp.org/dev/certs/ingress
+      property: pubcert
+    tls.key:
+      key: /corp.org/dev/certs/ingress
+      property: privkey
+
+  # used to fetch all properties from a secret.
+  # if multiple dataFrom are specified, secrets are merged in the specified order
+  dataFrom:
+  - key: /user/all-creds
+
+# status holds the timestamp and status of last last sync
+status:
+  lastSync: 2020-09-01T18:19:17.263Z # ISO 8601 date string
+  status: success # or failure
+
+```
+
+This API makes the options more explicit rather than having annotations.
+
+
+### External Secret Store
+
+The store configuration in an `ExternalSecret` may contain a lot of redundancy, this can be factored out into its own CRD.
+These stores are defined in a particular namespace using `SecretStore` **or** globally with `GlobalSecretStore`.
+
+```yaml
+apiVerson: external-secrets.k8s.io/v1alpha1
+kind: SecretStore # or ClusterSecretStore
+metadata:
+  name: vault
+  namespace: example-ns
+spec:
+
+  # optional.
+  # used to select the correct KES controller (think: ingress.ingressClassName)
+  # The KES controller is instantiated with a specific controller name
+  # and filters ES based on this property
+  controller: "dev"
+
+  store:
+    # store implementation
+    vault:
+      server: "https://vault.example.com"
+      path: secret/data
+      auth:
+        kubernetes:
+          path: kubernetes
+          role: example-role
+          secretRef:
+            name: vault-secret
```

Comment:
Should `SecretStore` have a status? W/ secret stores implementing some sort of liveliness check so you can validate a `SecretStore` is healthy before adding secrets?

Reply:
I think this would be useful but slightly challenging. Some SecretStore types may not have permissions to do inspection, for instance if Vault role isn't granted[ lookup-self](https://www.vaultproject.io/api-docs/auth/token#lookup-a-token-self) permissions then the SecretStore may report failed liveness but still have permission to fetch ExternalSecrets.

I think my recommendation would be to implement this with best effort and documenting the potential permissions store backends may need to support the liveness check.

Reply:
Agree, determining the `status` of an SecretStore is challenging. I think it would be great to have `Events` on the `SecretStore` CRD this way it is easier to observe if a single ES or multiple ES are having problems.

A basic `is ready` / `is running` could look like this:

```yaml

status:
  # * Pending: prerequisites missing / e.g. referenced secret containing credentials
  # * Running: all dependencies are met
  phase: Running
  conditions:
  - type: Ready
    status: "False"
    reason: "ErrorConfig"
    message: "Unable to assume role arn:xxxx"
    lastTransitionTime: "2019-08-12T12:33:02Z"
    lastHeartbeatTime: "2020-09-23T16:27:53Z"
```